hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -145,19 +223,109 @@ impl FlatGraphBuilder {
                     inn: rhs_ends.inn,
                 };
                 self.links.push(link_ends);
-                outer_ends
+                Ok(outer_ends)
             }
             Pipeline::Operator(operator) => {
                 let op_span = Some(operator.span());
                 let nid = self
                     .flat_graph
                     .insert_node(Node::Operator(operator), current_varname.cloned());
-                Ends {
+                Ok(Ends {
                     inn: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(nid))),
                     out: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(nid))),
+                })
+            }
+            Pipeline::Import(import) => {
+                // TODO: https://github.com/rust-lang/rfcs/pull/3200
+                // this would be way better...
+                let mut dir = self.macro_invocation_path.clone();
+                dir.pop();
+
+                let file_contents = std::fs::read_to_string(dir.join(import.filename.value()))
+                    .map_err(|e| {
+                        Error::new(
+                            import.filename.span(),
+                            format!("filename: {}, err: {e}", import.filename.value()),
+                        )
+                    })?;
+
+                // TODO: see also above, parse_str sets all the spans in the resulting parsed token stream to the parent macro invocation span.
+                // This means that any error inside the imported module will manifest as a giant red squiggly line under the parent hydroflow_syntax!{} call.
+                let statements = match syn::parse_str::<HfCode>(&file_contents) {
+                    Ok(code) => code,
+                    Err(err) => {
+                        self.diagnostics.push(err.clone().into());
+                        return syn::Result::Err(err);
+                    }
+                };
+
+                let flat_graph_builder = crate::graph::FlatGraphBuilder::from_hfmodule(statements);
+                let (flat_graph, _uses, diagnostics) = flat_graph_builder.build();
+                diagnostics
+                    .iter()
+                    .for_each(crate::diagnostic::Diagnostic::emit);
+
+                self.merge_in(flat_graph, import.span())
+            }
+        }
+    }
+
+    /// Merge one flatgraph into the current flatgraph
+    /// other must be a flatgraph and not be partitioned yet.
+    fn merge_in(&mut self, other: HydroflowGraph, parent_span: Span) -> syn::Result<Ends> {
```

Comment:
this doesn't seem to be able to return `Err`?

But could return `-> Result<Ends, Diagnostic>` instead of the asserts/panics maybe. Or just return `Ends`

Reply:
I think panic here is right, it should not be able to get into that state, if it does that is a hydroflow compiler bug. I changed it to just return Ends.